cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/mq/codec"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/util"
+	"go.uber.org/zap"
+)
+
+var _ Producer = (*kafkaProducer)(nil)
+
+// messageMetaData is used to store the callback function for the message.
+type messageMetaData struct {
+	callback eventsink.CallbackFunc
+}
+
+// kafkaProducer is used to send messages to kafka.
+type kafkaProducer struct {
+	// id indicates which processor (changefeed) this sink belongs to.
+	id model.ChangeFeedID
+	// role indicates what this sink is used for.
+	role util.Role
+	// We hold the client to make close operation faster.
+	// Please see the comment of Close().
+	client sarama.Client
+	// asyncProducer is used to send messages to kafka asynchronously.
+	asyncProducer sarama.AsyncProducer
+	// closedMu is used to protect `closed`.
+	// We need to ensure that closed producers are never written to.
+	closedMu sync.RWMutex
+	// closed is used to indicate whether the producer is closed.
+	// We also use it to guard against double closes.
+	closed bool
+	// closedChan is used to notify the run loop to exit.
+	closedChan chan struct{}
+	// failpointCh is used to inject failpoints to the run loop.
+	// Only used in test.
+	failpointCh chan error
+}
+
+func (k *kafkaProducer) AsyncSendMessage(
+	ctx context.Context, topic string,
+	partition int32, message *codec.MQMessage,
+) error {
+	// We have to hold the lock to avoid writing to a closed producer.
+	// Close may be blocked for a long time.
+	k.closedMu.RLock()
+	defer k.closedMu.RUnlock()
+
+	// If the producer is closed, we should skip the message.
+	if k.closed {
+		// We return the context error directly rather than any other error.
+		// Because if producer already closed, it means the context maybe is canceling or canceled.
+		// So we shouldn't return other errors to the caller to mess up the shutdown process.
+		return ctx.Err()
+	}
+	failpoint.Inject("KafkaSinkAsyncSendError", func() {
+		// simulate sending message to input channel successfully but flushing
+		// message to Kafka meets error
+		log.Info("failpoint error injected", zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		k.failpointCh <- errors.New("kafka sink injected error")
+		failpoint.Return(nil)
+	})
+
+	failpoint.Inject("SinkFlushDMLPanic", func() {
+		time.Sleep(time.Second)
+		log.Panic("SinkFlushDMLPanic",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+	})
+
+	msg := &sarama.ProducerMessage{
+		Topic:     topic,
+		Partition: partition,
+		Key:       sarama.StringEncoder(message.Key),
+		Value:     sarama.ByteEncoder(message.Value),
+		Metadata:  messageMetaData{callback: message.Callback},
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.asyncProducer.Input() <- msg:
```

Comment:
Is it possible for `k.asyncProducer.Input()` to be blocked by some error caused by the Kafka cluster?

Reply:
I think it might be time consuming, but it won't get stuck because after fan-in all the way, it will go to the broker producer's buffer.